docs/source/model/python_model.rst

```diff
@@ -0,0 +1,437 @@
+
+    # The same data validation works if you log and load the model as pyfunc
+    model_info = mlflow.pyfunc.log_model(
+        artifact_path="model",
+        python_model=model,
+        input_example=input_example,
+    )
+    pyfunc_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    print(pyfunc_model.predict(input_example))
+
+For callables, you can use @pyfunc decorator to enable data validation based on the type hints. 
+
+.. code-block:: python
+
+    from mlflow.pyfunc.utils import pyfunc
+
+
+    @pyfunc
+    def predict(model_input: list[Message]) -> list[str]:
+        return [msg.content for msg in model_input]
+
+
+    # The input_example can be a list of Message objects as defined in the type hint
+    input_example = [
+        Message(role="system", content="Hello"),
+        Message(role="user", content="Hi"),
+    ]
+    print(predict(input_example))  # Output: ['Hello', 'Hi']
+
+    # The input_example can also be a list of dict with the same schema as Message
+    input_example = [
+        {"role": "system", "content": "Hello"},
+        {"role": "user", "content": "Hi"},
+    ]
+    print(predict(input_example))  # Output: ['Hello', 'Hi']
+
+    # If your input doesn't match the schema, it will raise an exception
+    # e.g. passing a list of string here will raise an exception
+    predict(["hello"])
+    # Output: Failed to validate data against type hint `list[Message]`, invalid elements:
+    # [('hello', "Expecting example to be a dictionary or pydantic model instance for Pydantic type hint, got <class 'str'>")]
+
+.. note::
+    MLflow doesn't validate model output against the type hints, but the output type hint is used for model signature inference.
+
+
+Pydantic Model type hints data conversion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+For Pydantic model type hints, the input data can be either a Pydantic object or a dictionary that matches the schema of the Pydantic model.
+MLflow automatically converts the provided data to the type hint object before passing it to the predict function.
+Such as example in the last section: ``[{"role": "system", "content": "Hello"}]`` is converted to ``[Message(role="system", content="Hello")]`` inside the predict function.
+
+Below example demonstrates how to use a base class as type hint, while preserving fields in the subclasses.
+
+.. code-block:: python
+
+    from pydantic import BaseModel, ConfigDict
+    from mlflow.pyfunc.utils import pyfunc
+
+
+    class BaseMessage(BaseModel):
+        # set extra='allow' to allow extra fields in the subclass
+        model_config = ConfigDict(extra="allow")
+
+        role: str
+        content: str
+
+
+    class SystemMessage(BaseMessage):
+        system_prompt: str
+
+
+    class UserMessage(BaseMessage):
+        user_prompt: str
+
+
+    @pyfunc
+    def predict(model_input: list[BaseMessage]) -> list[str]:
+        result = []
+        for msg in model_input:
+            if hasattr(msg, "system_prompt"):
```

Comment:
do we need `hasattr`?

Reply:
Yes because UserMessage doesn't contain system_prompt field